File: lw7/MotorinaEvgeniya/lw7/lw7/ShellSorter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace lw7
{
    class ShellSorter
    {
        private List<int> m_array;

        private void Sort(int increment)
        {
            int j = 0;
            int temp = 0;
            for (int i = increment; i < m_array.Count(); i++)
            {
                temp = m_array[i];
                for (j = i; j >= increment && temp < m_array[j - increment]; j -= increment)
                {
                    m_array[j] = m_array[j - increment];
                }
                m_array[j] = temp;
            }
        }

        public void ShellSortWithThreads(List<int> array)
        {
            m_array = array;
            ThreadPool threadPool = new ThreadPool();
            for (int i = m_array.Count() / 2; i > 0; i /= 2)
            {
                threadPool.AddThread(Task.Factory.StartNew(() => Sort(i)));
            }
            threadPool.WaitThread();
            array = m_array;
        }

        public void ShellSort(List<int> array)
        {
            m_array = array;
            for (int i = m_array.Count() / 2; i > 0; i /= 2)
            {
                Sort(i);
            }
            array = m_array;
        }

    }
}

File: lw6/Motorina_Eugenia/lab6/lab6/ShellSorter.cpp
#include "stdafx.h"
#include "ThreadPool.h"
#include "ShellSorter.h"

CShellSorter::CShellSorter()
{
}

CShellSorter::~CShellSorter()
{
}

void CShellSorter::ShellSort(std::vector<int> & array)
{
	m_array = array;
	for (size_t i = m_array.size() / 2; i > 0; i /= 2)
	{
		Sort(i);
	}
	array = m_array;
}

void CShellSorter::ShellSortWithThreads(std::vector<int> & array)
{
	m_array = array;
	ThreadPool threadPool;
	for (size_t i = m_array.size() / 2; i > 0; i /= 2)
	{
		threadPool.AddThread(std::thread(boost::bind(&CShellSorter::Sort, this, i)));
	}
	threadPool.WaitThread();
	array = m_array;
}

void CShellSorter::Sort(size_t increment)
{
	size_t j = 0;
	int temp = 0;
	for (size_t i = increment; i < m_array.size(); i++)
	{
		temp = m_array[i];
		for (j = i; j >= increment && temp < m_array[j - increment]; j -= increment)
		{
			m_array[j] = m_array[j - increment];
		}
		m_array[j] = temp;
	}
}
